Add support for `shared-size` parameter for images queries

rvolosatovs opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**

Guard shared size computations with `shared-size` boolean URL parameter.
The reasoning for this change is to be able to query image shared size without having to rely on the more heavyweight `/system/df` endpoint.

**- How I did it**

Add an additional boolean parameter to `ImageService.Images` and add a URL parameter to the API endpoint mapping to it's value.
On a fast search over the codebase I did not see another URL parameter consisting of 2 words, so not sure about the naming convention, if there is one - went for kebab case.

**- How to verify it**

```sh
root@28f161c787d2:/go/src/github.com/docker/docker# docker build bundles -f bundles/ubuntu.Dockerfile 
Sending build context to Docker daemon  203.5MB
Step 1/2 : FROM ubuntu
 ---> 9873176a8ff5
Step 2/2 : RUN echo test > testfile
 ---> Running in 99dad2de2d7f
Removing intermediate container 99dad2de2d7f
 ---> 3036176f9a83
Successfully built 3036176f9a83
root@28f161c787d2:/go/src/github.com/docker/docker# curl --unix-socket /var/run/docker.sock http://localhost/images/json?shared-size=1 | jq .
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   998  100   998    0     0   974k      0 --:--:-- --:--:-- --:--:--  974k
[
  {
    "Containers": -1,
    "Created": 1624443646,
    "Id": "sha256:3036176f9a83226548a7b19a2d2bcc922945c5f43b3d591eec719e9dcb22e602",
    "Labels": null,
    "ParentId": "sha256:9873176a8ff5ac192ce4d7df8a403787558b9f3981a4c4d74afb3edceeda451c",
    "RepoDigests": [
      "<none>@<none>"
    ],
    "RepoTags": [
      "<none>:<none>"
    ],
    "SharedSize": 72742757,
    "Size": 72742762,
    "VirtualSize": 72742762
  },
  {
    "Containers": -1,
    "Created": 1623972689,
    "Id": "sha256:9873176a8ff5ac192ce4d7df8a403787558b9f3981a4c4d74afb3edceeda451c",
    "Labels": null,
    "ParentId": "",
    "RepoDigests": [
      "ubuntu@sha256:aba80b77e27148d99c034a987e7da3a287ed455390352663418c0f2ed40417fe"
    ],
    "RepoTags": [
      "ubuntu:latest"
    ],
    "SharedSize": 72742757,
    "Size": 72742757,
    "VirtualSize": 72742757
  }
] 
```

Also with filtering:
```sh
root@28f161c787d2:/go/src/github.com/docker/docker# curl --unix-socket /var/run/docker.sock 'http://localhost/images/json?shared-size=1&filters=%7B%22reference%22%3A%7B%22ubuntu%22%3Atrue%7D%7D'  | jq .
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   335  100   335    0     0   327k      0 --:--:-- --:--:-- --:--:--  327k
[
  {
    "Containers": -1,
    "Created": 1623972689,
    "Id": "sha256:9873176a8ff5ac192ce4d7df8a403787558b9f3981a4c4d74afb3edceeda451c",
    "Labels": null,
    "ParentId": "",
    "RepoDigests": [
      "ubuntu@sha256:aba80b77e27148d99c034a987e7da3a287ed455390352663418c0f2ed40417fe"
    ],
    "RepoTags": [
      "ubuntu:latest"
    ],
    "SharedSize": 72742757,
    "Size": 72742757,
    "VirtualSize": 72742757
  }
]
```

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

`/images` endpoint now supports boolean `shared-size` URL parameter, which enables shared size computation when set

thaJeztah commented:
Changing that boolean was my first idea, but (I think I left some notes on that on the internal ticket, but may have forgotten); we're mostly interested in the `SharedSize`; Setting the `withExtraAttrs` option will also collect the number of containers, which (at least at this point), I don't think we need. (Getting the list of containers may (potentially) have other implications, so if we don't need it (currently), we should probably avoid that)

Having a quick look, I think we should modify the logic in https://github.com/moby/moby/blob/1ba54a5fd0ba293db3bea46cd67604b593f2048b/daemon/images/images.go#L189-L218 to split the "size" part from the "container count" part.

Looking at that code, I think there's an issue; the code was written for `docker system df`, which will call the function _without_ any filters set; https://github.com/moby/moby/blob/1ba54a5fd0ba293db3bea46cd67604b593f2048b/daemon/images/images.go#L46

However, the `SharedSize` calculation is performed _after_ filtering (which works in the `docker system df` case, as it doesn't do any filtering), but it will calculate the _wrong_ shared sizes if the list _is_ filtered (it would only take other images in the result into account).

Even if we would put the calculation at the start of the loop; https://github.com/moby/moby/blob/1ba54a5fd0ba293db3bea46cd67604b593f2048b/daemon/images/images.go#L93


It looks like `allImages` is not always "all images"; https://github.com/moby/moby/blob/1ba54a5fd0ba293db3bea46cd67604b593f2048b/daemon/images/images.go#L64-L68

Commit:
API, daemon/images: add ImageListOptions and pass context

This makes it easier to add more options to the backend without having to change
the signature.

While we're changing the signature, also adding a context.Context, which is not
currently used, but probably should be at some point.

Signed-off-by: Roman Volosatovs <roman.volosatovs@docker.com>

rvolosatovs commented:
Updated the OP, added https://github.com/moby/moby/pull/42531/commits/200fab4b644eea54de7b337e3c699452a434f1dc , which is based on top of #42550 

@thaJeztah I suppose the safest way to handle this would be to make this feature opt-in, do you agree?

rvolosatovs commented:
Done in https://github.com/moby/moby/pull/42531/commits/0c51846574951248ad9212b653e3b2bf5177136d

thaJeztah commented on api/server/router/image/backend.go:
```diff
@@ -22,7 +22,7 @@ type Backend interface {
 	ImageDelete(imageRef string, force, prune bool) ([]types.ImageDeleteResponseItem, error)
 	ImageHistory(imageName string) ([]*image.HistoryResponseItem, error)
-	Images(imageFilters filters.Args, all bool, withExtraAttrs bool) ([]*types.ImageSummary, error)
+	Images(ctx context.Context, opts types.ImageListOptions, withContainerCount bool) ([]*types.ImageSummary, error)
```
LOL, not sure anymore on which PR I commented that; but given that we're adding a `ImageListOptions`, I think we should add the `withContainerCount` as field in that struct, and remove it as a separate boolean arg here

thaJeztah commented on api/server/router/image/image_routes.go:
```diff
@@ -235,7 +235,11 @@ func (s *imageRouter) getImagesJSON(ctx context.Context, w http.ResponseWriter,
 	}
 
-	images, err := s.backend.Images(imageFilters, httputils.BoolValue(r, "all"), false)
+	images, err := s.backend.Images(ctx, types.ImageListOptions{
```
The `shared-size` option should probably be gated by API version (and be ignored by `API < v1.41`

Perhaps here we should create a `opts := types.ImageListOptions{}` at the start of this function, then fill in options that are supported (possibly getting rid of the intermediate `imageFilters` variable)

rvolosatovs replied:
Something like this https://github.com/moby/moby/pull/42531/commits/7821e787813d356e018bd189e59a143b69e0516f ?

thaJeztah reviewed (changes requested):
Don't forget to also update https://github.com/moby/moby/blob/2bd46ed7e5a016f21b9259cec269340420fa7a63/api/swagger.yaml#L7185-L7192

And to add a mention of the new query-arg in https://github.com/moby/moby/blob/2bd46ed7e5a016f21b9259cec269340420fa7a63/docs/api/version-history.md

Commit:
API: add shared-size parameter to image queries

The reasoning for this change is to be able to query image shared size without having to rely on the more heavyweight `/system/df` endpoint.

Signed-off-by: Roman Volosatovs <roman.volosatovs@docker.com>

tianon commented:
I was having trouble understanding the motivation for this change (and feeling like I should already understand what's going on here) and discussed with @thaJeztah, and he provided the following explanation which helped me understand the intent much better:

> the system df command reuses the existing backend, but passes the "with extra attributes" argument, to calculate the shared size of images (and number of containers that use an image)
>
> The PR is to expose the shared size option on the /images/json (docker images / docker image ls) endpoint, so that the information can also be retrieved outside of docker system df

Can we add something simple like that to explain the motivation to the PR description/commit? :pray: :innocent:

rvolosatovs commented:
> I was having trouble understanding the motivation for this change (and feeling like I should already understand what's going on here) and discussed with @thaJeztah, and he provided the following explanation which helped me understand the intent much better:
> 
> > the system df command reuses the existing backend, but passes the "with extra attributes" argument, to calculate the shared size of images (and number of containers that use an image)
> > The PR is to expose the shared size option on the /images/json (docker images / docker image ls) endpoint, so that the information can also be retrieved outside of docker system df
> 
> Can we add something simple like that to explain the motivation to the PR description/commit? pray innocent

Added

> The reasoning for this change is to be able to query image shared size without having to rely on the more heavyweight `/system/df` endpoint.

to the OP, please let me know if you think this is not enough and whether we should record this in git history as well, for example.

thaJeztah commented:
Adding to git history, as part of the commit message is always good to have.

That said, this PR likely will need a rebase after the other PR is merged, so it's ok to do it at that point

> On 28 Jun 2021, at 18:34, Roman Volosatovs ***@***.***> wrote:
> 
> ﻿
> I was having trouble understanding the motivation for this change (and feeling like I should already understand what's going on here) and discussed with @thaJeztah, and he provided the following explanation which helped me understand the intent much better:
> 
> the system df command reuses the existing backend, but passes the "with extra attributes" argument, to calculate the shared size of images (and number of containers that use an image)
> The PR is to expose the shared size option on the /images/json (docker images / docker image ls) endpoint, so that the information can also be retrieved outside of docker system df
> 
> Can we add something simple like that to explain the motivation to the PR description/commit? pray innocent
> 
> Added
> 
> The reasoning for this change is to be able to query image shared size without having to rely on the more heavyweight /system/df endpoint.
> 
> to the OP, please let me know if you think this is not enough and whether we should record this in git history as well, for example.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

thaJeztah commented on api/server/router/image/image_routes.go:
```diff
@@ -222,20 +222,23 @@ func (s *imageRouter) getImagesJSON(ctx context.Context, w http.ResponseWriter,
-		filterParam := r.Form.Get("filter")
-		if filterParam != "" {
-			imageFilters.Add("reference", filterParam)
+	if versions.LessThan(httputils.VersionFromContext(ctx), "1.41") {
+		if s := r.Form.Get("filter"); s != "" {
+			opts.Filters = filters.NewArgs(filters.Arg("reference", s))
+		}
+	} else {
+		filters, err := filters.FromJSON(r.Form.Get("filters"))
+		if err != nil {
+			return err
 		}
```
This doesn't look correct; in the old code;

1. the "filters" option would be applied
2. if API < 1.41, and the `filter` (singular) option was set, added to the list of filters as `reference` filter

I think `2.` was because before `--filter` was added, it was possible to filter images by name (reference) only as a positional argument (`docker images <search>`); when the `--filter` flag was added, code was added to merge the existing (positional) filter argument with the new `--filter` options. Iterating over the "blame" view history, I see the `filter` option was implemented at least 6 years ago; https://github.com/moby/moby/blame/23750fb80280e6770590b0ea30781c43f42e430d/api/server/image.go#L367, well before API v1.41 (looks like it was originally added in https://github.com/moby/moby/commit/5f3812ec97f9a951bceb482fdcd355b6cbc1b113, as part of https://github.com/moby/moby/pull/4430

At the API level, the old `filter` query parameter was deprecated in favor of `filters` parameter (equivalent to `--filter reference=xxx`), see https://github.com/moby/moby/pull/27872

I'm a bit confused now, as deprecated the `filter` query-arg (for API v1.28 and above), but I see I later opened https://github.com/moby/moby/pull/40491 to deprecate it _again_. When I opened that PR, it looks like the API v1.28 check was removed (??), so I used API v1.41 instead. Will have to find when that v1.28 check was removed (as that would've "un-deprecated" the option?

thaJeztah replied:
Ah; found where it was removed; https://github.com/moby/moby/pull/31844

Reason at the time was that it was "to be deprecated after two releases", and it was yet unknown what API version that would be, so the check was replaced by a "TODO" instead, which wasn't picked up until v1.41

rvolosatovs replied:
The logic was changed due to your comment here https://github.com/moby/moby/pull/42531#discussion_r659820126, as we discussed offline.
I will drop https://github.com/moby/moby/pull/42531/commits/7821e787813d356e018bd189e59a143b69e0516f, such API changes can be handled in another PR, where such changes are in scope.

thaJeztah commented on daemon/disk_usage.go:
```diff
@@ -26,7 +26,11 @@ func (daemon *Daemon) SystemDiskUsage(ctx context.Context) (*types.DiskUsage, er
 	// Get all top images with extra attributes
-	allImages, err := daemon.imageService.Images(filters.NewArgs(), false, true)
+	allImages, err := daemon.imageService.Images(ctx, types.ImageListOptions{
+		Filters:        filters.NewArgs(),
```
Not for this PR, but wondering if we should change `Filters` to be a pointer, so that it can be left `nil` if no filtering is used

thaJeztah commented:
Note: this should also update the default API version; we can merge https://github.com/moby/moby/pull/42063 before this once review progresses (I'll rebase that one, so that you can possibly include that in this PR; we can merge it separately or as part of this one (I somewhat prefer separately to make it easier to find back, but not either way would be fine)

thaJeztah commented on api/swagger.yaml:
```diff
@@ -7202,6 +7202,11 @@ paths:
             - `reference`=(`<image-name>[:<tag>]`)
             - `since`=(`<image-name>[:<tag>]`,  `<image id>` or `<image@digest>`)
           type: "string"
+        - name: "shared-size"
```
The API and docs changes should go into the same commit as the code changes (as one shouldn't get in without the other).

thaJeztah commented on api/server/router/image/image_routes.go:
```diff
@@ -226,16 +226,21 @@ func (s *imageRouter) getImagesJSON(ctx context.Context, w http.ResponseWriter,
 			imageFilters.Add("reference", filterParam)
 		}
+	} else {
+		sharedSize = httputils.BoolValue(r, "shared-size")
```
It's somewhat "obvious", but perhaps good to add a comment here, a-la;

```suggestion
		// support for the "shared-size" parameter was added in API 1.42
		sharedSize = httputils.BoolValue(r, "shared-size")
```

thaJeztah commented on api/server/router/image/image_routes.go:
```diff
@@ -226,16 +226,21 @@ func (s *imageRouter) getImagesJSON(ctx context.Context, w http.ResponseWriter,
 	}
 
-	images, err := s.backend.Images(imageFilters, httputils.BoolValue(r, "all"), false)
+	images, err := s.backend.Images(ctx, types.ImageListOptions{
```
for the first "refactor" commit message that adds the `ImageListOptions` here, I think "support more control" is a bit confusing, because in that commit it's not yet adding "more control". Perhaps something like:

```
API, daemon/images: add ImageListOptions and pass context

This makes it easier to add more options to the backend without having to change
the signature.

While we're changing the signature, also adding a context.Context, which is not
currently used, but probably should be at some point.
```

thaJeztah commented:
Weird; Swagger validation is for some reason failing on a non-related change; 

```
[2021-06-30T09:28:44.409Z] The swagger spec at "api/swagger.yaml" is invalid against swagger specification 2.0. see errors :
[2021-06-30T09:28:44.409Z] - definitions.NetworkSettings.properties.SecondaryIPAddresses.description in body must be of type string: "null"
[2021-06-30T09:28:44.409Z] - definitions.NetworkSettings.properties.SecondaryIPv6Addresses.description in body must be of type string: "null"
```

Looks like it complains about the empty description, but no clue why it fails on this PR (and not on other ones); https://github.com/moby/moby/blob/7b9275c0da707b030e62c96b679a976f31f929d3/api/swagger.yaml#L1304-L1318

thaJeztah commented:
opened https://github.com/moby/moby/pull/42583 to update the swagger files 🤷‍♂️

rvolosatovs commented:
> Weird; Swagger validation is for some reason failing on a non-related change;
> 
> ```
> [2021-06-30T09:28:44.409Z] The swagger spec at "api/swagger.yaml" is invalid against swagger specification 2.0. see errors :
> [2021-06-30T09:28:44.409Z] - definitions.NetworkSettings.properties.SecondaryIPAddresses.description in body must be of type string: "null"
> [2021-06-30T09:28:44.409Z] - definitions.NetworkSettings.properties.SecondaryIPv6Addresses.description in body must be of type string: "null"
> ```
> 
> Looks like it complains about the empty description, but no clue why it fails on this PR (and not on other ones);
> 
> https://github.com/moby/moby/blob/7b9275c0da707b030e62c96b679a976f31f929d3/api/swagger.yaml#L1304-L1318

Yes, saw that as well, but assumed that is a known issue!

thaJeztah commented on api/types/client.go:
```diff
@@ -235,10 +235,21 @@ type ImageImportOptions struct {
+	All bool
+
+	// Filters is a JSON-encoded set of filter arguments.
+	// filter is a shell glob string applied to repository names.
```
`filter is a shell glob string applied to repository names.` looks to be a description for the legacy `?filter` parameter. Probably should be removed here, as this describes the "filters" as a whole (not that parameter)

We can move that comment to api/server/router/image/image_routes.go to describe the property there, in case it's useful.

thaJeztah commented on api/server/router/image/image_routes.go:
```diff
@@ -226,18 +226,21 @@ func (s *imageRouter) getImagesJSON(ctx context.Context, w http.ResponseWriter,
 		if filterParam != "" {
 			imageFilters.Add("reference", filterParam)
 		}
+	} else {
+		// NOTE: Support for the "shared-size" parameter was added in API 1.42.
```
nit: _slightly_ leaning toward making this separate from the check above, i.e.

```suggestion
	}

	if versions.GreaterThanOrEqualTo(httputils.VersionFromContext(ctx), "1.42") {
		// Support for the "shared-size" parameter was added in API 1.42.	
```

thaJeztah commented on api/swagger.yaml:
```diff
@@ -7202,6 +7202,11 @@ paths:
           type: "string"
+        - name: "shared-size"
+          in: "query"
+          description: "Show shared size as a `SharedSize` field on each image."
```
(thinking out loud) wondering if we should mention `compute` here and/or some mention that setting this option will introduce additional overhead. I see we're not doing this (currently) for some other `size` parameter, but maybe we should.

rvolosatovs replied:
I also think `compute` is a better term here, but I went with "show" for consistency.
Perhaps we should revise all of these in a separate PR?

thaJeztah reviewed (commented):
Overall looks good (left some thoughts/feedback inline, but no strong "blockers").

If we go ahead with this PR, I suggest we merge https://github.com/moby/moby/pull/42063 first (separately), to make the API version update more easily discoverable (then rebase this PR)

thaJeztah commented:
`TestClientWithRequestTimeout` failure is unrelated; flaky test will hopefully be fixed by https://github.com/moby/moby/pull/42580


```
=== Failed
=== FAIL: pkg/plugins TestClientWithRequestTimeout (0.00s)
    client_test.go:254: assertion failed: expected an error, got nil: expected error
```

Failure on Windows is `TestNetworkDBIslands` - also a flaky test

Let me kick CI

rvolosatovs commented:
Blocked by https://github.com/moby/moby/pull/42063 (will rebase and rerequest review once it is merged)

thaJeztah reviewed (approved):
LGTM

thaJeztah commented:
@tianon @tonistiigi PTAL
